zerver/tests/test_channel_permissions.py

```diff
@@ -746,3 +748,241 @@ def test_change_stream_message_retention_days_requires_realm_owner(self) -> None
+        self.unsubscribe(polonius, stream.name)
+
+        # If stream is private, test which permissions require having
+        # content access to the channel. We do not test this for
+        # public streams since all non-guest users have content
+        # access to them.
+        if stream.invite_only:
+            # Unsubscribe user from private stream to test gaining
+            # content access from group settings.
+            self.unsubscribe(iago, stream.name)
+            self.unsubscribe(shiva, stream.name)
+            self.unsubscribe(hamlet, stream.name)
+            do_change_stream_group_based_setting(
+                stream, "can_administer_channel_group", members_group, acting_user=iago
+            )
+
+            if (
+                property_name
+                not in Stream.stream_permission_group_settings_requiring_content_access
+                and property_name != "invite_only"
+            ):
+                # Users without content access can modify properties not in
+                # stream_permission_group_settings_requiring_content_access
+                # can change privacy of non-private streams, and can archive
+                # or unarchive them.
+                check_channel_property_update(iago)
+                check_channel_property_update(shiva)
+                check_channel_property_update(hamlet)
+            else:
+                error_msg = "Channel content access is required."
+                # Even realm and channel admins need content access to a private
+                # channel for changing its privacy and to update the permissions
+                # in stream_permission_group_settings_requiring_content_access.
+                check_channel_property_update(iago, allow_fail=True, error_msg=error_msg)
+                check_channel_property_update(shiva, allow_fail=True, error_msg=error_msg)
+                check_channel_property_update(hamlet, allow_fail=True, error_msg=error_msg)
+
+                do_change_stream_group_based_setting(
+                    stream, "can_add_subscribers_group", moderators_group, acting_user=iago
+                )
+                check_channel_property_update(hamlet, allow_fail=True, error_msg=error_msg)
+                check_channel_property_update(iago)
+                check_channel_property_update(shiva)
+
+                do_change_stream_group_based_setting(
+                    stream, "can_subscribe_group", hamletcharacters_group, acting_user=iago
+                )
+                check_channel_property_update(hamlet)
+
+                # Reset the setting values to "Nobody" group.
+                do_change_stream_group_based_setting(
+                    stream, "can_add_subscribers_group", nobody_group, acting_user=iago
+                )
+                do_change_stream_group_based_setting(
+                    stream, "can_subscribe_group", nobody_group, acting_user=iago
+                )
```

Comment:
Can we extract the private stream testing in it's own function somehow. Even if it's used once, it  will make it easier to read the whole thing I think.

Reply:
Do you mean we need to add a function in the same scope as `check_channel_property_update`? I don't think doing that helps in readability and defining it outside would mean passing too many parameters.